### PR TITLE
Add shard closing listener

### PR DESCRIPTION
**Pull request by henningandersen**
After #108145, the after shard closed listener is no longer called on the cluster state applier thread. Introduce another event that is called on the applier thread.

Relates #108145

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java [hunk 1]
```diff
@@ -55,6 +55,11 @@ default void afterIndexShardStarted(IndexShard indexShard) {}
      */
     default void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {}
 
+    /**
+     * Called after the index shard has been marked closed. It could still be waiting for the async close of the engine.
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java:59]: nit Might be good to extend the comment to say that afterIndexShardClosed may be called earlier (due to asynchronicity in rare cases).

**Review by kingherc — APPROVED** [review]
LGTM, one nit

**Review by fcofdez — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java:59]: Sure, done in 1217103fe087580889fe69e00307670dd4d56239

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Pull request merged**